### ISSUE TITLE
Exported SavedModel has a huge size

### ISSUE THREAD

**Issue by ktobah** [ultralytics/ultralytics#20018]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

I trained an OBB model and I would like to export it to SavedModel format to serve it using TF serving. I need to have a dynamic batch size. At first, I exported using the following args and got a model with a size of around 260MB. But the batch size is 1:

`yolo export model=weights/best.pt format=saved_model keras=True data=dataset/data.yaml imgsz=800 nms=True iou=0.6 conf=0.5`

I tried to search how to have a dynamic batch size, and seems the only way is to set a fixed value using batch, so I did this and ended up with an almost 2GB model (the .pb file alone)!

`yolo export model=weights/best.pt format=saved_model keras=True data=dataset/data.yaml imgsz=800 nms=True iou=0.6 conf=0.5 batch=64`
 
I am using the docker image from Utralytics. Why such a large size? And is there a way to have a dynamic batch size (-1, None) rather than a fixed value?

### Environment

```
Ultralytics 8.3.93 🚀 Python-3.11.10 torch-2.5.1+cu124 CPU (Intel Xeon Gold 6146 3.20GHz)
YOLO11x-obb summary (fused): 199 layers, 58,736,758 parameters, 0 gradients, 202.7 GFLOPs
```

### Minimal Reproducible Example

`yolo export model=weights/best.pt format=saved_model keras=True data=dataset/data.yaml imgsz=800 nms=True iou=0.6 conf=0.5 batch=64`

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ktobah, thank you for bringing this to our attention 🚀! We appreciate your detailed report and your interest in using Ultralytics.

For issues like this, we recommend checking the [Docs](https://docs.ultralytics.com/) for additional insights on exporting models, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions about export formats and configurations may already be addressed there.

Since this appears to be a 🐛 Bug Report, providing a more detailed [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) would help us investigate further. For instance, sharing:
- A snippet of your dataset or configuration file
- A clear explanation of the steps taken before encountering the issue

Additionally, for specific export-related concerns, verifying that you're using the latest version of the `ultralytics` package is essential. You can upgrade with the following command:

```bash
pip install -U ultralytics
```

## Community and Support

Feel free to connect with the Ultralytics community for real-time discussions and support:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://reddit.com/r/Ultralytics)

## Verified Environments

We recommend running YOLO in one of these verified environments to ensure compatibility:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. These tests aim to ensure the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across different platforms.

This is an automated response to assist you faster. An Ultralytics engineer will follow up soon to provide further assistance. Stay tuned! 😊

**Comment by ktobah** [comment]
I reduced the batch size to 8, and I can confirm the final size depends on the batch size, which is not an expected behavior:

```
Ultralytics 8.3.93 🚀 Python-3.11.10 torch-2.5.1+cu124 CPU (Intel Xeon Gold 6146 3.20GHz)
YOLO11x-obb summary (fused): 199 layers, 58,736,758 parameters, 0 gradients, 202.7 GFLOPs

PyTorch: starting from 'weights/best.pt' with input shape (8, 3, 800, 800) BCHW and output shape(s) (8, 300, 7) (112.7 MB)
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1743782728.016919       7 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1743782728.027412       7 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1743782728.049652       7 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1743782728.049674       7 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1743782728.049677       7 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1743782728.049680       7 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
requirements: Ultralytics requirement ['protobuf>=5'] not found, attempting AutoUpdate...
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
wandb 0.19.8 requires protobuf!=4.21.0,!=5.28.0,<6,>=3.19.0; python_version > "3.9" and sys_platform == "linux", but you have protobuf 6.30.2 which is incompatible.
tflite-support 0.4.4 requires protobuf<4,>=3.18.0, but you have protobuf 6.30.2 which is incompatible.
tensorflow 2.19.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<6.0.0dev,>=3.20.3, but you have protobuf 6.30.2 which is incompatible.
ydf 0.11.0 requires protobuf<6.0.0,>=5.29.1, but you have protobuf 6.30.2 which is incompatible.
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable.It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.
Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com
Collecting protobuf>=5
  Downloading protobuf-6.30.2-cp39-abi3-manylinux2014_x86_64.whl.metadata (593 bytes)
Downloading protobuf-6.30.2-cp39-abi3-manylinux2014_x86_64.whl (316 kB)
Installing collected packages: protobuf
  Attempting uninstall: protobuf
    Found existing installation: protobuf 3.20.3
    Uninstalling protobuf-3.20.3:
      Successfully uninstalled protobuf-3.20.3
Successfully installed protobuf-6.30.2

requirements: AutoUpdate success ✅ 12.2s, installed 1 package: ['protobuf>=5']
requirements: ⚠️ Restart runtime or rerun command for updates to take effect


TensorFlow SavedModel: starting export with tensorflow 2.19.0...

ONNX: starting export with onnx 1.17.0 opset 19...
ONNX: slimming with onnxslim 0.1.48...
ONNX: export success ✅ 53.0s, saved as 'weights/best.onnx' (233.2 MB)
TensorFlow SavedModel: starting TFLite export with onnx2tf 1.27.0...
I0000 00:00:1743785199.193900       7 devices.cc:67] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1743785199.194184       7 single_machine.cc:374] Starting new session
W0000 00:00:1743785224.825787       7 tf_tfl_flatbuffer_helpers.cc:365] Ignored output_format.
W0000 00:00:1743785224.825831       7 tf_tfl_flatbuffer_helpers.cc:368] Ignored drop_control_dependency.
I0000 00:00:1743785250.279488       7 devices.cc:67] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0
I0000 00:00:1743785250.279749       7 single_machine.cc:374] Starting new session
W0000 00:00:1743785274.936238       7 tf_tfl_flatbuffer_helpers.cc:365] Ignored output_format.
W0000 00:00:1743785274.936275       7 tf_tfl_flatbuffer_helpers.cc:368] Ignored drop_control_dependency.
TensorFlow SavedModel: export success ✅ 2581.2s, saved as 'weights/best_saved_model' (648.6 MB)

Export complete (2598.7s)
Results saved to /usr/src/app/weights
Predict:         yolo predict task=obb model=weights/best_saved_model imgsz=800  
Validate:        yolo val task=obb model=weights/best_saved_model imgsz=800 data=/usr/src/app/dataset/data.yaml  
Visualize:       https://netron.app
💡 Learn more at https://docs.ultralytics.com/modes/export
```

**Comment by ktobah** [comment]
Another thing, with the above exported model, I cannot run prediction. I get the following error even if I use a single picture (which worked fine before with the batch size = 1):

```
    result_future = stub.Predict(request, 120)
  File "/usr/local/lib/python3.9/site-packages/grpc/_channel.py", line 1181, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.9/site-packages/grpc/_channel.py", line 1006, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "2 root error(s) found.
  (0) INVALID_ARGUMENT: Input to reshape is a tensor with 480000 values, but the requested shape has 3840000
         [[{{function_node __inference_model_314_layer_call_and_return_conditional_losses_15980}}{{node tf.reshape_2/Reshape}}]]
         [[PartitionedCall/_71]]
  (1) INVALID_ARGUMENT: Input to reshape is a tensor with 480000 values, but the requested shape has 3840000
         [[{{function_node __inference_model_314_layer_call_and_return_conditional_losses_15980}}{{node tf.reshape_2/Reshape}}]]
0 successful operations.
0 derived errors ignored."
        debug_error_string = "UNKNOWN:Error received from peer ipv4:172.7.0.2:8500 {grpc_message:"2 root error(s) found.\n  (0) INVALID_ARGUMENT: Input to reshape is a tensor with 480000 values, but the requested shape has 3840000\n\t [[{{function_node __inference_model_314_layer_call_and_return_conditional_losses_15980}}{{node tf.reshape_2/Reshape}}]]\n\t [[PartitionedCall/_71]]\n  (1) INVALID_ARGUMENT: Input to reshape is a tensor with 480000 values, but the requested shape has 3840000\n\t [[{{function_node __inference_model_314_layer_call_and_return_conditional_losses_15980}}{{node tf.reshape_2/Reshape}}]]\n0 successful operations.\n0 derived errors ignored.", grpc_status:3, created_time:"2025-04-04T17:10:39.683811466+00:00"}"
```

The model definition is (I just renamed the inputs and outputs):

```
The given SavedModel SignatureDef contains the following input(s):
  inputs['input_image'] tensor_info:
      dtype: DT_FLOAT
      shape: (8, 800, 800, 3)
      name: serving_default_input_image:0
The given SavedModel SignatureDef contains the following output(s):
  outputs['output_0'] tensor_info:
      dtype: DT_FLOAT
      shape: (8, 300, 7)
      name: PartitionedCall:0
Method name is: tensorflow/serving/predict
```

And the definition of the model that works is the following:

```
The given SavedModel SignatureDef contains the following input(s):
  inputs['input_image'] tensor_info:
      dtype: DT_FLOAT
      shape: (1, 800, 800, 3)
      name: serving_default_input_image:0
The given SavedModel SignatureDef contains the following output(s):
  outputs['output_0'] tensor_info:
      dtype: DT_FLOAT
      shape: (1, 300, 7)
      name: PartitionedCall:0
Method name is: tensorflow/serving/predict
```

**Comment by Y-T-G** [comment]
That's normal behavior. SavedModel doesn't support dynamic batch sizes. Only fixed batch sizes. And the size is also normal for FP32 export. If you want smaller size, you should use either `half` or `int8`.

**Comment by ktobah** [comment]
In yolov5, I was used to export to SavedModel with a None (-1) batch size and it worked fine.

**Comment by glenn-jocher** [comment]
Thank you for reaching out! For YOLO11 exports to TensorFlow SavedModel format, dynamic batch sizes aren't currently supported due to framework limitations. The `batch` argument sets a fixed batch dimension as shown in the Export documentation. This is expected behavior and differs from previous YOLOv5 implementations. 

For dynamic batch support, we recommend exporting to ONNX format with `dynamic=True`, which allows flexible batch dimensions. You can achieve this with:

```bash
yolo export model=

**Comment by Y-T-G** [comment]
YOLOv5 was natively rewritten in TensorFlow and export is performed with native model. Ultralytics uses `onnx2tf`. Dynamic export doesn't work.

**Comment by ktobah** [comment]
Alright, thank you for your explanation.

**Comment by glenn-jocher** [comment]
You're welcome! For YOLO11 SavedModel exports, dynamic batch sizes aren't supported due to TensorFlow conversion limitations, but you can achieve dynamic batches by exporting to ONNX format with `dynamic=True`. This allows flexible batch dimensions while maintaining performance. For implementation details, see the ONNX Export documentation. Let us know if you need further guidance! 🚀
